### PR TITLE
Fix amp tests

### PR DESCRIPTION
**Pull request by neggert**
A few changes to fix AMP-related problems in our tests:

* With `opt_level="O1"` (the default), AMP patches many torch functions, which breaks any tests that run afterwards. This patch introduces a pytest extension that lets tests be marked with `@pytest.mark.spawn` so that they are run in their own process using `torch.multiprocessing.spawn` so that the main python interpreter stays un-patched.

* Since AMP defaults to O1 now, DP tests no longer throw exceptions. Remove the `pytest.raises` that expects the exception.

* Since AMP patches torch functions, CPU inference no longer works.
Skip prediction step for AMP tests.

Note that there are still a couple of unrelated test failures that will be addressed in another PR.

### CODE REVIEW

**Comment by neggert** [comment]
Travis failure appears to be due to this unrelated issue: https://github.com/pytorch/vision/issues/1718

File: tests/test_amp.py [hunk 1]
```diff
@@ -124,26 +128,6 @@ def test_amp_gpu_ddp_slurm_managed(tmpdir):
     assert trainer.resolve_root_node_address('abc[23-24]') == 'abc23'
     assert trainer.resolve_root_node_address('abc[23-24, 45-40, 40]') == 'abc23'
 
-    # test model loading with a map_location
```
Comment (Borda) [inline tests/test_amp.py]: maybe just comment it?

**Comment by Borda** [comment]
tests are failing on PIL, I found somewhere that adding `Pollow>=4.2` solves the issue

- https://github.com/python-pillow/Pillow/issues/4130
- https://pillow.readthedocs.io/en/stable/releasenotes/7.0.0.html#pillow-version-constant

**Comment by neggert** [comment]
Supposedly torchvision will do a bug fix release early next week (0.4.2?) that will resolve the Pillow issue. The fix is already in torchvision master. I’m inclined to just wait for that and update our minimum torchvision version.

On Jan 4, 2020, at 9:40 AM, Jirka Borovec <notifications@github.com> wrote:

﻿

tests are failing on PIL, I found somewhere that adding Pollow>=4.2 solves the issue

  *   python-pillow/Pillow#4130<https://github.com/python-pillow/Pillow/issues/4130>
  *   https://pillow.readthedocs.io/en/stable/releasenotes/7.0.0.html#pillow-version-constant

**Comment by hugovk** [comment]
[torchvision v0.5.0](https://github.com/pytorch/vision/releases/tag/v0.5.0) has been released with the fix:

1. Require `torchvision>=0.5.0`
2. If Pillow was temporarily pinned, remove the pin

**Comment by Borda** [comment]
but `torchvision>=0.5.0` has some JIT issues... #694

**Pull request merged**